src/main/java/cuchaz/enigma/gui/GuiController.java

```diff
@@ -603,4 +620,50 @@ public void setDecompiler(DecompilerService service) {
 		decompiler = createDecompiler();
 		refreshCurrentClass(null, RefreshMode.FULL);
 	}
+
+	public EnigmaClient getClient() {
+	    return client;
+    }
+
+    public EnigmaServer getServer() {
+	    return server;
+    }
+
+    public void createClient(String username, String ip, int port) throws IOException {
+		client = new EnigmaClient(this, ip, port);
+		client.connect();
+		client.sendPacket(new LoginC2SPacket(project.getJarChecksum(), username));
+	}
+
+    public void createServer(int port) throws IOException {
+		server = new EnigmaServer(project.getJarChecksum(), EntryRemapper.mapped(project.getJarIndex(), new HashEntryTree<>(project.getMapper().getObfToDeobf())), port);
```

Comment:
So we create a full server on the client. 

For cases like just editing stuff on my own machine, would it make more sense to possible have an abstraction that directly forwards the packets locally to the client (Like how the integrated servers are connected to locally via netty) rather than forward the data to a port and then have to query for it, possibly leaving latency

Reply:
There is no server unless you explicitly start one. And even once there is a server there shouldn't be any noticeable latency - it changes the client-side mappings directly then updates the server with those changes, then on the odd occasion the server doesn't like your changes, it sends a packet back to you to revert them.